src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/AuditUtility.cs

```diff
@@ -0,0 +1,287 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Common;
+using NuGet.LibraryModel;
+using NuGet.Packaging.Core;
+using NuGet.Protocol.Model;
+using NuGet.Versioning;
+
+namespace NuGet.Commands.Restore.Utility
+{
+    internal struct AuditUtility
+    {
+        private readonly ProjectModel.RestoreAuditProperties _restoreAuditProperties;
+        private readonly string _projectFullPath;
+        private readonly IEnumerable<RestoreTargetGraph> _targetGraphs;
+        private readonly IReadOnlyList<IVulnerabilityInformationProvider> _vulnerabilityInfoProviders;
+        private readonly ILogger _logger;
+
+        public AuditUtility(
+            ProjectModel.RestoreAuditProperties restoreAuditProperties,
+            string projectFullPath,
+            IEnumerable<RestoreTargetGraph> graphs,
+            IReadOnlyList<IVulnerabilityInformationProvider> vulnerabilityInformationProviders,
+            ILogger logger)
+        {
```

Comment:
Should any of these inputs be checked for `null`?  It looks like only our code will new it up but those inputs might come from a 3rd party caller.

Reply:
The struct is `internal`, so it's not (yet?) a public API. If it was a public API, then yes, definitely. However, I think that `RestoreTargetGraph` is so specific to restore that there isn't a benefit to making it public for other developers to use in their own apps. So ultimately, I chose to be conservative, because it's easy to add APIs in the future, but it's hard to remove APIs when we make bug fixes or add new features.